### ISSUE TITLE
Async version of `list` command?

### ISSUE THREAD

**Issue by sseemayer** [durch/rust-s3#47]
I'm trying to use `rust-s3` in a method returning a Future to asynchronously get data from an S3 bucket and am currently using version `0.16.0-alpha2`. Using `get_object_async` is already working fine for me, however, I would now like to list the contents of the Bucket to download further objects in the same async function. However, if I call `mybucket.list`, it seems that the list method is not currently compatible with asynchronous operation:

```
executor failed to spawn task: tokio::spawn failed (is a tokio runtime running this future?)
```

Would it be possible to add a `list_async` function similar to the `get_object_async` method?

Many thanks for providing this very useful library!

**Comment by durch** [comment]
@sseemayer I'm glad you found this useful! I took a quick look at the code. For legacy reasons `list` seems to be using a slightly different code path, I need to unify that anyway, which should get you the async version pretty much for free. I'll keep you posted :)

**Comment by durch** [comment]
@sseemayer I did a bunch of breaking changes, not all of them related to this issue, version `0.18.0-beta` adds `Bucket` methods `list_all_async` and `list_page_async`  which should get you what you need. Please also note that the returned types of all `*_async` commands are now just a single `future`, which should be more pleasant to use. Let me know if it works out :)

**Comment by sseemayer** [comment]
This looks excellent, thank you very much! I almost got my use case working already but am still debugging some compile errors on my end.

Context: I am using rust-s3 within an actix-web application. I am writing a moderately-complex route handler that will pull together information from an S3 bucket consisting of several objects, process them and return views into the processed data. I'm still doing something wrong with the `Send` marker trait not getting passed on when using `join_all` to join the futures for getting individual objects and me needing the `Send` trait for actix.

**Comment by durch** [comment]
For sure, very happy to help, always interested in more usage :). I’d be happy to take a look at the code your working on, maybe I can help, if the code is something you can share that is :)

**Comment by sseemayer** [comment]
OK, I've tried now for a while but am not sure if there is a way to get this to work without altering `rust-s3`. Here is my minimal example:

```rust
fn get_bucket_list<'a>(
    bucket: &'a s3::bucket::Bucket,
    path: &'a str,
) -> Box<dyn Future<Item = Vec<String>, Error = Error> + Send> {
    let fut = bucket
        .list_all_async(path, Some("/"))
        .map(|res| {
            res.into_iter()
                .flat_map(|a| a.contents.into_iter().map(|o| o.key.clone()))
                .collect()
        })
        .map_err(|_e| format_err!("Error listing bucket"));

    Box::new(fut)
}
```
I need to return a `Future` that is also `Send` because I want to return it to an actix-web route handler. The code above refuses to compile with the following longish error message:

```
error[E0277]: `<impl futures::future::Future as futures::future::Future>::Error` cannot be sent between threads safely
   --> src/source.rs:139:5
    |
139 |     Box::new(fut)
    |     ^^^^^^^^^^^^^ `<impl futures::future::Future as futures::future::Future>::Error` cannot be sent between threads safely
    |
    = help: within `futures::future::or_else::OrElse<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, std::result::Result<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`, the trait `std::marker::Send` is not implemented for `<impl futures::future::Future as futures::future::Future>::Error`
    = note: required because it appears within the type `std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>`
    = note: required because it appears within the type `std::option::Option<std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>>`
    = note: required because it appears within the type `futures::future::result_::FutureResult<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>`
    = note: required because it appears within the type `futures::future::chain::Chain<futures::future::loop_fn::LoopFn<futures::future::and_then::AndThen<impl futures::future::Future, std::result::Result<futures::future::loop_fn::Loop<(std::option::Option<&str>, std::vec::Vec<s3::serde_types::ListBucketResult>), (std::option::Option<std::string::String>, std::vec::Vec<s3::serde_types::ListBucketResult>)>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:84 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]::{{closure}}[0]) 0:std::vec::Vec<s3::serde_types::ListBucketResult>]>, [closure@DefId(166:83 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]) 0:&s3::bucket::Bucket, 1:&str, 2:std::option::Option<&str>]>, futures::future::result_::FutureResult<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:85 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[1])]>`
    = note: required because it appears within the type `futures::future::and_then::AndThen<futures::future::loop_fn::LoopFn<futures::future::and_then::AndThen<impl futures::future::Future, std::result::Result<futures::future::loop_fn::Loop<(std::option::Option<&str>, std::vec::Vec<s3::serde_types::ListBucketResult>), (std::option::Option<std::string::String>, std::vec::Vec<s3::serde_types::ListBucketResult>)>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:84 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]::{{closure}}[0]) 0:std::vec::Vec<s3::serde_types::ListBucketResult>]>, [closure@DefId(166:83 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[0]) 0:&s3::bucket::Bucket, 1:&str, 2:std::option::Option<&str>]>, std::result::Result<std::vec::Vec<s3::serde_types::ListBucketResult>, <impl futures::future::Future as futures::future::Future>::Error>, [closure@DefId(166:85 ~ s3[9bed]::bucket[0]::{{impl}}[0]::list_all_async[0]::{{closure}}[1])]>`
    = note: required because it appears within the type `impl futures::future::Future`
    = note: required because it appears within the type `futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>`
    = note: required because it appears within the type `futures::future::chain::Chain<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, futures::future::result_::FutureResult<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`
    = note: required because it appears within the type `futures::future::or_else::OrElse<futures::future::map::Map<impl futures::future::Future, [closure@src/source.rs:132:14: 136:10]>, std::result::Result<std::vec::Vec<std::string::String>, failure::error::Error>, [closure@src/source.rs:137:18: 137:37]>`
    = note: required for the cast to the object type `dyn futures::future::Future<Error = failure::error::Error, Item = std::vec::Vec<std::string::String>> + std::marker::Send`
```

if it's any use, `Error` is `failure::Error`. It looks like some `Error` associated type is not `Send` making the rest of the `Future` not `Send`.

**Comment by durch** [comment]
Taking a look, lets see how far I can get just adding `Send` to return futures

**Comment by durch** [comment]
@sseemayer got it to work, I've had to fiddle around with errors a bit, definitely need to take another pass at that code, and likely get rid of `snafu`. It is also a bit over-specified at the moment, but it should get you what you need. Let me know how it goes :)

version `0.18.0-beta1` pushed to crates.io

**Comment by sseemayer** [comment]
After spending some hours fighting with rustc and conflicting lifetime errors in a Futures context, I am giving up for today and will spend some more time tomorrow to come up with a minimal example to share.

**Comment by sseemayer** [comment]
Here is [a minimal project combining actix-web and rust-s3](https://github.com/sseemayer/actix-rust-s3-experiment). Note that there is an `index_list` route handler commented out that encompasses my best efforts of getting a Future response based on `bucket.list_all_async` working.

If I understand correctly, the problem is that `list_all_async` will return a future whose lifetime is the same as the passed bucket, and I need to dynamically create buckets in the route handler from user-provided parameters in my "real" use case. I haven't figured out how to pass a future that owns the Bucket or any other way to make rustc happy about my lifetimes.

A small side comment on usability: I noticed that the associated `Error` type of the `Future`s returned by `rust-s3` don't seem to `impl Error` (or `Debug` or `Display` for that matter), so I cannot easily pass on a backtrace.

**Comment by durch** [comment]
Fudged with lifetimes a bit, peppered clones all over the place, not very happy about that, but it'll do for now. Published `0.18.0-beta2`. 

You're right about the lifetime issue, as it was defined explicitly in the function definition, I've just defined it as an anonymous lifetime on the return value, so it should solve the issue. 

You're dead on about the `Error`, I just threw it in there for now, but I'll whip it into order for the non-beta release.

**Comment by sseemayer** [comment]
I figured out how to get it working with a small change - see #49 and the working example https://github.com/sseemayer/actix-rust-s3-experiment

**Comment by durch** [comment]
@sseemayer thank you! Very happy we got this to behave :) ! Published `0.18.0-beta3` to crates.io. I'll do some cleanup and get this out of beta in the coming weeks.

**Comment by sseemayer** [comment]
Thanks to you! I'm closing the issue since this is all that I needed :)
